numba/npyufunc/parallel.py

```diff
@@ -317,22 +329,56 @@ def threading_layer():
         return _threading_layer
 
 
+def _check_tbb_version_compatible():
```

Comment:
If I explicitly set the threading layer to TBB:

```python
numba.config.THREADING_LAYER="tbb"
```

then this works OK for preventing me loading an old TBB, and prints out the message explaining the required version and the version it found, but then produces a long stacktrace and a generic message about not being able to load a threading layer:

```
In [16]: @njit(parallel=True) 
    ...: def vecadd(c, b, a): 
    ...:     for i in prange(len(c)): 
    ...:         c[i] = a[i] + b[i] 
    ...:                                                                                                                                                     

In [17]: vecadd(c, b, a)                                                                                                                                     
/home/gmarkall/numbadev/numba/numba/npyufunc/parallel.py:358: NumbaWarning: The TBB threading layer requires TBB version 2019 update 5 or later i.e. TBB_INTERFACE_VERSION >= 11005. Found TBB_INTERFACE_VERSION = 11004. The TBB threading layer is disabled.
  warnings.warn(problem)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/numbadev/numba/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    716     try:
--> 717         yield
    718     except NumbaError as e:

# ... lots of frames snipped ...

~/numbadev/numba/numba/npyufunc/parallel.py in raise_with_hint(required)
    432                     hint = hintmsg % ("One of:\n%s" % options)
--> 433                 raise ValueError(errmsg % hint)
    434 

ValueError: No threading layer could be loaded.
HINT:
Intel TBB is required, try:
$ conda/pip install tbb

During handling of the above exception, another exception occurred:

LoweringError                             Traceback (most recent call last)
<ipython-input-17-8620bb091d65> in <module>
----> 1 vecadd(c, b, a)

# .. lots of frames snipped ...

~/numbadev/numba/numba/six.py in reraise(tp, value, tb)
    667         if value.__traceback__ is not tb:
    668             raise value.with_traceback(tb)
--> 669         raise value
    670 
    671 else:

LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
No threading layer could be loaded.
HINT:
Intel TBB is required, try:
$ conda/pip install tbb

File "<ipython-input-16-e59b5d768e80>", line 3:
def vecadd(c, b, a):
    for i in prange(len(c)):
    ^

[1] During: lowering "id=0[LoopNest(index_variable = parfor_index.20, range = (0, c_size0.15, 1))]{18: <ir.Block at <ipython-input-16-e59b5d768e80> (3)>}Var(parfor_index.20, <ipython-input-16-e59b5d768e80>:3)" at <ipython-input-16-e59b5d768e80> (3)
```

The later messages make it sound like TBB just wasn't found, and it would be easy to miss the first message about the version in the noise - I'm not sure how easy it is to change so that the last error message printed is about the TBB version, but that would probably help avoid some user error.

Reply:
Hmm. For me, it prints the warning but then does the computation anyway

```py
$ cat test-tbb.py
from numba import njit, prange
import numpy as np
a, b, c = [np.array([1]) for i in range(3)]
@njit(parallel=True)
def vecadd(c, b, a):
    for i in prange(len(c)):
        c[i] = a[i] + b[i]
vecadd(c, b, a)

print(c)

from numba import config

print(config.THREADING_LAYER)
```

```
$ NUMBA_THREADING_LAYER=tbb python test-tbb.py
/Users/aaronmeurer/Documents/numba/numba/npyufunc/parallel.py:358: NumbaWarning: The TBB threading layer requires TBB version 2019 update 5 or later i.e. TBB_INTERFACE_VERSION >= 11005. Found TBB_INTERFACE_VERSION = 11004. The TBB threading layer is disabled.
  warnings.warn(problem)
[2]
tbb
$ python test-tbb.py
/Users/aaronmeurer/Documents/numba/numba/npyufunc/parallel.py:358: NumbaWarning: The TBB threading layer requires TBB version 2019 update 5 or later i.e. TBB_INTERFACE_VERSION >= 11005. Found TBB_INTERFACE_VERSION = 11004. The TBB threading layer is disabled.
  warnings.warn(problem)
[2]
default
```

Presumably if it is default it should fallback, and if it is set to force tbb the warning should be an error. 

The error you showed is what I get if I remove tbb entirely and try to force it. IMO the main problem here is that the error is way too verbose (IPython is partly to blame for that too) and doesn't have the real message at the very bottom, but then again, that's par for the course with the current state of error messages in numba.

Reply:
I also just tested it and can confirm that numba fails correctly with old tbb at build time.

Reply:
hmmm, I suspect this in part comes from that fall-over behaviour, where by in the case of asking for e.g. a "thread safe" backend, the selector will try TBB then OpenMP. This https://github.com/numba/numba/pull/4615/files#diff-ec5b6bd5010b2b32f80834aad542033cR354 probably needs to raise an import error so that if there is the wrong version present it won't be loaded. With that error raising being added this:
```python
import numpy as np
import numba
numba.config.THREADING_LAYER="threadsafe"
from numba import njit, prange


a, b, c = [np.array([1]) for i in range(3)]
@njit(parallel=True)
def vecadd(c, b, a):
    for i in prange(len(c)):
        c[i] = a[i] + b[i]
vecadd(c, b, a)

print(numba.threading_layer())
```
produces "omp" for me on the last line when I deliberately install an incompatible TBB version, it also produces the warning about TBB being incompatible.

Reply:
> I also just tested it and can confirm that numba fails correctly with old tbb at build time.

Great, thanks.

Reply:
Fixed in https://github.com/numba/numba/pull/5044/commits/a82bc3189df9c50629ae25e71efa037ed886d142